src/esp/scene/HM3DSemanticScene.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "HM3DSemanticScene.h"
+#include "SemanticScene.h"
+
+#include <Corrade/Utility/FormatStl.h>
+#include <fstream>
+#include <map>
+#include <sstream>
+#include <string>
+
+namespace Cr = Corrade;
+namespace esp {
+namespace scene {
+
+bool SemanticScene::loadHM3DHouse(
+    const std::string& houseFilename,
+    SemanticScene& scene,
+    const quatf& rotation /* = quatf::FromTwoVectors(-vec3f::UnitZ(),
+                                                       geo::ESP_GRAVITY) */ ) {
+  if (!checkFileExists(houseFilename, "loadHM3DHouse")) {
+    return false;
+  }
+  // open stream and determine house format version
+  std::ifstream ifs = std::ifstream(houseFilename);
+  std::string header;
+  std::getline(ifs, header);
+  if (header != "HM3D Semantic Annotations") {
+    ESP_ERROR() << "Unsupported HM3D House format header" << header
+                << "in file name" << houseFilename;
+    return false;
+  }
+
+  return buildHM3DHouse(ifs, scene, rotation);
+}  // SemanticScene::loadHM3DHouse
+
+namespace {
+
+// These structs are here to make constructing the actual Semantic objects
+// easier, due to the format of the HM3D ssd files.
+
+struct TempHM3DObject {
+  int objInstanceID;
+  std::string categoryName;
+  std::string objInstanceName;
+  Mn::Vector3ub color;
+  std::shared_ptr<SemanticRegion> region;
+
+  void setObjInstanceAndName(int _objInstanceID) {
+    objInstanceID = _objInstanceID;
+    objInstanceName =
+        Cr::Utility::formatString("{}_{}", categoryName, objInstanceID);
+  }
+};
+
+struct TempHM3DRegion {
+  int regionID;
+  std::vector<TempHM3DObject*> objInstances;
+};
+struct TempHM3DCategory {
+  int ID;
+  std::string name;
+  std::vector<TempHM3DObject*> objInstances;
+
+  std::shared_ptr<HM3DObjectCategory> category_;
+};
+
+}  // namespace
+
+bool SemanticScene::buildHM3DHouse(std::ifstream& ifs,
+                                   SemanticScene& scene,
+                                   const quatf& /*rotation*/) {
+  // convert a hex string containing 3 bytes to a vector3ub (represents a color)
+  auto getVec3ub = [](const std::string& hexValStr) -> Mn::Vector3ub {
+    size_t len = hexValStr.length();
+    std::vector<uint8_t> resVec;
+    for (size_t i = 0; i < len; i += 2) {
+      uint8_t x = strtoul(hexValStr.substr(i, 2).c_str(), nullptr, 16);
+      resVec.push_back(x);
+    }
+    return Mn::Vector3ub{resVec[0], resVec[1], resVec[2]};
```

Comment:
Uhh, C++ string-to-number APIs are *bad*, but this is somehow even worse :sweat_smile: Why not parse the whole thing as a single integer and then use that?

```suggestion
    const unsigned val = std::stoul(hexValStr, nullptr, 16);
    return {uint8_t((val >> 16) & 0xff),
            uint8_t((val >>  8) & 0xff),
            uint8_t((val >>  0) & 0xff)};
```

Reply:
That's what I get for trying to work out an SO suggestion.